Implement the BlockingQueue as a wrapper around the QueueFile

pschichtel opened the pull request:
The resulting BlockingFileQueue is thread-safe and unbounded.
Thread-safety is implemented with a single lock around all operations against the backing queue.

Closes #198

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

The resulting BlockingFileQueue is thread-safe and unbounded.
Thread-safety is implemented with a single lock around all operations against the backing queue.

f2prateek commented on tape/src/main/java/com/squareup/tape2/BlockingFileQueue.java:
```diff
@@ -0,0 +1,407 @@
+ * This is a simple and unbounded @{@link BlockingQueue} implementation as a wrapper around a @{@link QueueFile}.
+ * Thread safety is implemented using a single lock around all operations to the backing @{@link QueueFile}.
+ */
+public class BlockingFileQueue implements BlockingQueue<byte[]> {
```
Rather than `byte[]` - what do you think about using a converter interface https://github.com/square/tape/blob/master/tape/src/main/java/com/squareup/tape2/ObjectQueue.java#L86-L92 to allow custom types to be used?

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

Switched to an ObjectQueue<T> as the backing store and delegate closable.

pschichtel commented:
I switched to the ObjectQueue as the backing structure as I generally agree that it's more useful, however I'm not entirely sure what overhead this adds in case I'm really only using byte[] buffers (as I am in the project where code this originally comes from).

pschichtel commented:
The current implementation is also unbounded and I don't intend to implement an upper bound because the backing queue also doesn't have an upper bound.
If an upper bound is added to the QueueFile in the future it would be good to add one here as well.

pschichtel commented:
I'll take a look into directly accessing QueueFile for BlockingQueue<byte[]> instance without duplicating the whole class.

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

QueueFile is now an ObjectQueue<byte[]> and can thus be used without overhead in the blocking queue. ObjectQueue's file():QueueFile method has been moved to FileObjectQueue where it fits better as InMemoryObjectQueue doesn't have a backing file.
Removed the ByteArrayConverter again as it was the initial attempt to optimize for byte[] queues.

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

sneakyfy the IO exception as done in other places.

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

Increased coverage and minor cleanup

pschichtel commented:
@f2prateek So that's the state I would go with. In order to implement `ObjectQueue<byte[]>` on QueueFile I had to break the current API a little as `file():QueueFile` on `ObjectQueue` conflicted with `file():File` in `QueueFile`, so I moved `file():QueueFile` to `FileObjectQueue`. This way we don't have the noop `file():QueueFile` on `InMemoryObjectQueue` and `QueueFile` can implement `ObjectQueue<byte[]>` basically as-is.

pschichtel commented on tape/src/main/java/com/squareup/tape2/QueueFile.java:
```diff
@@ -591,7 +590,7 @@ public int size() {
    * @throws NoSuchElementException if the queue is empty
    */
-  public void remove() throws IOException {
+  @Override public void remove() throws IOException {
```
this method could be removed, as it is exactly the same es in `ObjectQueue`. Should I go ahead and remove it or keep the diff minimal?

pschichtel commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,448 @@
+      E peek = queue.peek();
+      if (peek == null) {
+        // this won't happen unless the backing queue has been shared.
+        throw new IllegalStateException("backing queue empty");
```
should probably be a `NoSuchElementException` as well.

pschichtel commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,448 @@
+  @Override public E remove() {
+    lock.lock();
+    try {
+      if (queue.isEmpty()) {
```
could be changed to a null check on `peek()`'s result

NightlyNexus replied:
the isEmpty check seems clearer to me. 👍

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

minor cleanup

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+ *
+ * @param <E> the element type
+ */
+public class BlockingObjectQueue<E> implements BlockingQueue<E>, Closeable {
```
final

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+
+  private final ObjectQueue<E> queue;
+
+  public BlockingObjectQueue(ObjectQueue<E> queue) {
```
how about just having the static factory?

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+   * @param <T> the element type
+   * @return a BlockObjectQueue implementation
+   */
+  public static <T> BlockingObjectQueue<T> create(QueueFile qf, ObjectQueue.Converter<T> conv) {
```
maybe just take in an `ObjectQueue<T>`?

pschichtel replied:
These are convenience methods, I'd rather remove the static factories than the constructor. I did them similar to the one in ObjectQueue.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+   * @param qf the queue file which should not be shared to other places
+   * @return a BlockObjectQueue implementation
+   */
+  public static BlockingObjectQueue<byte[]> create(QueueFile qf) {
```
ditto above. maybe can remove this.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      return true;
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
+      return false;
```
how about `throw QueueFile.<Error>getSneakyThrowable(e);`? then we don't need the unreachable return statement.

pschichtel replied:
Throwing outside the method doesn't work (I tried it, do you know a way that works?).

NightlyNexus replied:
hm, should work. there are some existing usages. https://github.com/square/tape/blob/445cd3fd0a7b3ec48c9ea3e0e86663fe6d3735d8/tape/src/main/java/com/squareup/tape2/FileObjectQueue.java#L97

pschichtel replied:
It must have been late... :D

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      queue.remove();
+      return peek;
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
```
throws

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      E peek = queue.peek();
+      if (peek == null) {
+        // this won't happen unless the backing queue has been shared.
+        throw new NoSuchElementException();
```
`ConcurrentModificationException`? unsure.

pschichtel replied:
good idea.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      return peek;
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
+      return null;
```
throws

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+  }
+
+  /**
+   * The underlying {@link QueueFile} only supports removing the head, so this will only work if
```
underlying ObjectQueue.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      }
+      Iterator<E> it = queue.iterator();
+      while (it.hasNext()) {
+        if (Objects.deepEquals(it.next(), o)) {
```
yeah

NightlyNexus replied:
also, we may need to reimplement the array check, since Tape has traditionally needed to work back to JDK 6.

pschichtel replied:
I'd prefer to not do that if it would be acceptable to raise the jvm target.

pschichtel replied:
`Objects.deepEquals()` is available since JDK 1.7 which is also already EOL, so I personally don't think this is problematic.

NightlyNexus replied:
Android. Some folks still target devices before KitKat, and the build toolchains don't rewrite everything.

I'm personally in favor of dropping support for 6, though.

pschichtel replied:
Isn't most of the 1.7 stuff available via the compat libs? Not sure

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      Iterator<E> it = queue.iterator();
+      while (it.hasNext()) {
+        c.add(it.next());
+        it.remove();
```
👍

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+    }
+  }
+
+  @Override public int drainTo(Collection<? super E> c, int maxElements) {
```
nit: should we remove some code and have the other drainTo overload call this drainTo? `drainTo(c, queue.size())`?

pschichtel replied:
agree.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      return peek;
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
+      return null;
```
throws `QueueFile.<Error>getSneakyThrowable(e);`

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      return peek;
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
+      return null;
```
throws

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      return queue.peek();
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
+      return null;
```
throws

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+    lock.lock();
+    try {
+      if (queue.isEmpty()) {
+        return null;
```
here, it could be simply `return queue.peek();` (avoiding the redundant isEmpty check).

pschichtel replied:
agreed

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      }
+      return queue.peek();
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
```
throws

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+/**
+ * This is a simple and unbounded {@link BlockingQueue} implementation as a wrapper
+ * around a {@link QueueFile}. Thread safety is implemented using a single lock around all
+ * operations to the backing {@link QueueFile}.
```
around and backing an ObjectQueue

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+  }
+
+  /**
+   * The underlying {@link QueueFile} only supports removing the head, so this will only work if
```
ObjectQueue

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      E peek = queue.peek();
+      if (peek == null) {
+        // this won't happen unless the backing queue has been shared.
+        throw new NoSuchElementException();
```
unsure if it should be a ConcurrentModifificationException.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      return peek;
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
+      return null;
```
throws

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+  @Override public Iterator<E> iterator() {
+    lock.lock();
+    try {
+      return queue.iterator();
```
is locking while creating this factory necessary?
the iterator still iterates on the backing ObjectQueue (not a clone of it or anything).

pschichtel replied:
probably doesn't make a difference.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+    lock.lock();
+    try {
+      for (Object e : c) {
+        if (!contains(e)) {
```
won't this be a problem? the lock won't ever get acquired in `contains` because it's already held here.

pschichtel replied:
it's a ReentrantLock, from its docs: `The method will return immediately if the current thread already owns the lock.`

NightlyNexus replied:
👍

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      return true;
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
+      return false;
```
throws

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+    boolean changed = false;
+    try {
+      for (Object e : c) {
+        if (remove(e)) {
```
looks like a deadlock here, too.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+  }
+
+  /**
+   * The underlying {@link QueueFile} only supports removing the head, so this will only work if
```
ObjectQueue. also, the head matching comment seems a little out of place on retainAll.

pschichtel replied:
minimal example that will throw:
```
queue = [a, b]
retainCollection = [a]
```

this would *not* `.remove()` on the head, but on the second element.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+      if (queue.isEmpty()) {
+        return false;
+      } else {
+        clear();
```
deadlock?

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/BlockingObjectQueue.java:
```diff
@@ -0,0 +1,450 @@
+    try {
+      queue.clear();
+    } catch (IOException e) {
+      QueueFile.<Error>getSneakyThrowable(e);
```
nit: throws for consistency and clarity.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/ObjectQueue.java:
```diff
@@ -25,9 +26,6 @@
   }
 
-  /** The underlying {@link QueueFile} backing this queue, or null if it's only in memory. */
-  public abstract @Nullable QueueFile file();
```
this has value. let's leave it.
if we do make QueueFile an ObjectQueue, we could rename this API to `queueFile()`.

pschichtel replied:
with `file():QueueFile` in `ObjectQueue` you would need to null-check on the result to be save, now with the method in `FileObjectQueue` you have to check for the implementation and are then guaranteed to get a `QueueFile` back, not a huge difference.
I think this could be improved by making the factory methods return the specific implementations.

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/QueueFile.java:
```diff
@@ -424,11 +423,6 @@ private long remainingBytes() {
   }
 
-  /** Returns true if this queue contains no entries. */
-  public boolean isEmpty() {
```
let's leave this.

pschichtel replied:
It's inherited from `ObjectQueue`

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/QueueFile.java:
```diff
@@ -582,25 +576,16 @@ private void checkForComodification() {
-   *
-   * @throws NoSuchElementException if the queue is empty
-   */
-  public void remove() throws IOException {
```
and leave this. people use QueueFile directly, and the convenience APIs are convenient!

pschichtel replied:
It's inherited from `ObjectQueue`

NightlyNexus commented on tape/src/main/java/com/squareup/tape2/QueueFile.java:
```diff
@@ -52,7 +51,7 @@
  * @author Bob Lee (bob@squareup.com)
  */
-public final class QueueFile implements Closeable, Iterable<byte[]> {
+public final class QueueFile extends ObjectQueue<byte[]> {
```
hmmmm, could you get the same functionality by making a FileObjectQueue<byte[]>? a little roundabout, but it keeps QueueFile focused.

pschichtel replied:
I don't see the benefit, can you elaborate?

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

more cleanup and a new test

pschichtel committed:
Closes #198: Implement the BlockingQueue as a wrapper around the QueueFile

directly throw the sneaky exceptions

pschichtel commented:
Any updates here?

pschichtel commented:
@NightlyNexus I guess this won't be merged?
